vpr/src/base/region.cpp

```diff
@@ -0,0 +1,81 @@
+#include "region.h"
+
+/// @brief sentinel value for indicating that a subtile has not been specified
+constexpr int NO_SUBTILE = -1;
+
+Region::Region() {
+    sub_tile = NO_SUBTILE;
+
+    //default rect for a region is (-1, -1, -1, -1)
+    region_bounds.set_xmin(-1);
+    region_bounds.set_ymin(-1);
+    region_bounds.set_xmax(-1);
+    region_bounds.set_ymax(-1);
+}
+
+vtr::Rect<int> Region::get_region_rect() {
+    return region_bounds;
+}
+
+void Region::set_region_rect(int _xmin, int _ymin, int _xmax, int _ymax) {
+    region_bounds.set_xmin(_xmin);
+    region_bounds.set_xmax(_xmax);
+    region_bounds.set_ymin(_ymin);
+    region_bounds.set_ymax(_ymax);
+}
+
+int Region::get_sub_tile() {
+    return sub_tile;
+}
+
+void Region::set_sub_tile(int _sub_tile) {
+    sub_tile = _sub_tile;
+}
+
+bool Region::do_regions_intersect(Region region) {
+    bool intersect = true;
+
+    vtr::Rect<int> region_rect = region.get_region_rect();
+    vtr::Rect<int> intersect_rect;
+
+    intersect_rect = intersection(region_bounds, region_rect);
+
+    /**
+     * if the intersection rectangle is empty or the subtile of the two regions does not match,
+     * the regions do not intersect
+     */
+    if (intersect_rect.empty() || sub_tile != region.get_sub_tile()) {
+        return intersect = false;
+    }
+
+    return intersect;
+}
+
+Region Region::regions_intersection(Region region) {
```

Comment:
Rename to intersection, and make it a global function at the bottom of the header file (outside class):
Region intersection (Region a, Region b);
// May have to be a friend function of class Region.